src/charm.py

```diff
@@ -303,9 +307,12 @@ def _is_unit_waiting_to_join_cluster(self) -> bool:
         return (
             self.unit.get_container(CONTAINER_NAME).can_connect()
             and self.unit_peer_data.get("member-state") == "waiting"
-            and self._mysql.is_data_dir_initialised()
-            and not self.unit_peer_data.get("unit-initialized")
-            and int(self.app_peer_data.get("units-added-to-cluster", 0)) > 0
+            and self.unit_configured
+            and (
+                not self.unit_initialized
+                or not self._mysql.is_instance_in_cluster(self.unit_label)
```

Comment:
we test for instance in cluster here due cluster metadata existing for the unit on scale-up after scale-down, i.e. storage reuse

Reply:
The unit can be correctly in the cluster but offline in a transient state when mysqld is starting and initializing the group replication plugin. At this stage the unit will try to rejoin the cluster. This can take some time and running the mysqlsh api join command fails.